ddl/dm_schema_tracker.go

```diff
@@ -0,0 +1,686 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright 2013 The ql Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSES/QL-LICENSE file.
+
+package ddl
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ngaut/pools"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/ddl/util"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/meta/autoid"
+	"github.com/pingcap/tidb/owner"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/parser/charset"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/statistics/handle"
+	"github.com/pingcap/tidb/table"
+	pumpcli "github.com/pingcap/tidb/tidb-binlog/pump_client"
+	"github.com/pingcap/tidb/util/collate"
+	"github.com/pingcap/tidb/util/dbterror"
+)
+
+var _ DDL = SchemaTracker{}
+
+// SchemaTracker is used to track schema changes by DM. It implements DDL interface and by applying DDL, it updates the
+// table structure to keep tracked with upstream changes.
+type SchemaTracker struct {
+	*infoschema.InfoStore
+}
+
+// NewSchemaTracker creates a SchemaTracker. lowerCaseTableNames has the same meaning as MySQL variable lower_case_table_names.
+func NewSchemaTracker(lowerCaseTableNames int) SchemaTracker {
+	return SchemaTracker{
+		InfoStore: infoschema.NewInfoStore(lowerCaseTableNames),
+	}
+}
+
+// CreateSchema implements the DDL interface.
+func (d SchemaTracker) CreateSchema(ctx sessionctx.Context, stmt *ast.CreateDatabaseStmt) error {
```

Comment:
How are we going to add unit tests to this file?

Reply:
`Checker` can compare the result DBInfo of `ddl` and `SchemaTracker`. I added `Checker` to some unit tests in executor/ddl_test.go . Please help me find more suitable unit test to add `Checker`!